src/riak_kv_index_hashtree.erl

```diff
@@ -190,6 +193,21 @@ get_lock(Tree, Type) ->
 get_lock(Tree, Type, Pid) ->
     gen_server:call(Tree, {get_lock, Type, Pid}, infinity).
 
+
+%% @doc Acquire the lock for the specified index_hashtree if not already
+%%      locked, and associate the lock with the calling process.
+%%      When not_built pid are saved and given the lock when build finish.
+-spec wait_for_lock(pid(), any()) -> ok | building | not_built | already_locked.
+wait_for_lock(Tree, Type) ->
+    wait_for_lock(Tree, Type, self()).
+
+%% @doc Acquire the lock for the specified index_hashtree if not already
+%%      locked, and associate the lock with the provided pid.
+-spec wait_for_lock(pid(), any(), pid()) -> ok | not_built | already_locked.
+wait_for_lock(Tree, Type, Pid) ->
+    gen_server:call(Tree, {wait_for_lock, Type, Pid}, 60000).
```

Comment:
Should we also set up a monitor to detect if the tree process dies and therefore forgets that this process is to be notified when the trees are built?  I guess it could be done by the caller